File: Lista03_HerancaEPolimorfismo/Heranca/app/src/main/java/Lista03/Estoque.java
package Lista03;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Estoque {
    List<Produto> listaProdutos = new ArrayList<Produto>();

    public void cadastrarProduto(Produto p){
        this.listaProdutos.add(p);
        return;
    }

    public void retirarEstoque(Produto p){
        Produto c = this.consultarProduto(p.getCodigo());

        if (c == null){
            System.out.println("Produto nao encontrado");
            return;
        }

        this.listaProdutos.remove(c);

        return;
    }

    public Produto consultarProduto(int codigo){

        for (Produto p : listaProdutos){
            if (p.getCodigo() == codigo) return p;
        }

        return null;
    } 

    public double custoTotal(){
        double precoTotal = 0.0;
        
        for (Produto p : listaProdutos){
            precoTotal += p.getPreco_unitario();
        }

        return precoTotal;
    }

    public void imprimirTodosProdutos(){
        for (Produto p : listaProdutos){
	        System.out.println("Codigo: "+p.codigo); 
	        System.out.println("Descricao: "+p.descricao); 
	        System.out.println("Quantidade em estoque: "+p.quantidade_estoque); 
	        System.out.println("======================");
        }
    }

}

File: Lista11_Python/Ex04_Heranca/estoque.py
from produto import *
from produto_perecivel import *
from produto_perecivel_especial import *

class Estoque:
    arrayProdutos = []

    def ImprimirTodosProdutos(self):

        print("----------------")
        for produto in self.arrayProdutos:
            print(produto)
            print("----------------")

    def CadastrarProduto(self, produto):

        def ConsultaProdutoInterno(codigo):
            for produto in self.arrayProdutos:
                if produto._codigo == codigo:
                    return False

            return True

        if (issubclass(type(produto), Produto) or type(produto) == Produto):
            codigo = produto.getCodigo()
            consulta = ConsultaProdutoInterno(codigo)
            if (consulta == True):
                self.arrayProdutos.append(produto)
                return True
            else:
                print("Código de produto repitido!")
                return False

        print("Produto inválido!")
        return False

    def ConsultarProduto(self, codigo):
        
        for produto in self.arrayProdutos:
            if produto._codigo == codigo:
                print("Produto de código "+str(codigo)+" encontrado!")
                print(produto)
                return produto

        print("Produto de código "+str(codigo)+" NÃO encontrado!")

        return None


    def RetirarProduto(self, produto):

        def ConsultaProdutoInterno(codigo):
            for produto in self.arrayProdutos:
                if produto._codigo == codigo:
                    return produto

            return None

        if (issubclass(type(produto), Produto) or type(produto) == Produto):
            codigo = produto.getCodigo()
            consulta = ConsultaProdutoInterno(codigo)
            if consulta != None:
                self.arrayProdutos.remove(produto)
                return
            else:
                print("Produto inexistente")
            return

        print("Produto inválido!")
        return None

    def RetirarProdutoCodigo(self, codigo):

        def ConsultaProdutoInterno(codigo):
            for produto in self.arrayProdutos:
                if produto._codigo == codigo:
                    return produto
            return None

        consulta = ConsultaProdutoInterno(codigo)
        if consulta != None:
            self.arrayProdutos.remove(consulta)
            return
        else:
            print("Produto inexistente")
            return 

        print("Produto inválido!")
        return



    def CustoTotal(self):

        custoTotal = 0
        for produto in self.arrayProdutos:
            custoTotal += produto._precoUnitario

        print("Custo total: "+str(custoTotal))

        return custoTotal
